exercises/collatz-conjecture/example.rkt

```diff
@@ -0,0 +1,23 @@
+#lang racket
+
+(provide collatz)
+
+; uses a hash table to speed up subsequent calls
+(define (collatz-length ht n)
+  (cond 
+    [(hash-ref ht n #f) => (lambda (colval) colval)]
+    [else
+     (define colval 0)
+     (if (even? n) 
+         (set! colval (add1 (collatz-length ht (/ n 2))))
+         (set! colval (add1 (collatz-length ht (add1 (* 3 n))))))
+     (hash-set! ht n colval)
+     colval]))
+
+(define (collatz n)
+  (define h (make-hash (list '(1 . 0))))
```

Comment:
What's the general consensus on ad-hoc approaches like this vs something like the following?

```racket
(require memoize)

(define/memo (collatz-length n)
  ...)
```

Reply:
Hmmm, yes, using memoize is better... in fact I think in this case it isn't justified... I guess overthought the problem a litte... thinking that you might call collatz-length multiple times.  I'm gonna simplify this a lot.